### PR TITLE
Improved OIDC compliance

### PR DESCRIPTION
**Pull request by luisrudge**
### Description

This update improves the SDK support for OpenID Connect. In particular, it modifies the sign in verification phase by substituting backchannel based checks with id_token validation.

### Testing

- [x] This change adds test coverage for new/changed/fixed functionality

### CODE REVIEW

File: src/jwt.ts [hunk 1]
```diff
@@ -53,37 +58,134 @@ export const decode = (token: string) => {
...
+        );
+      }
+    }
+  } else if (decoded.claims.aud !== options.aud) {
+    throw new Error(
+      `Audience (aud) claim mismatch in the ID token; expected "${options.aud}" but found "${decoded.claims.aud}"`
+    );
+  }
+
+  if (!decoded.claims.nonce) {
```
Comment (lbalmaceda) [inline src/jwt.ts]: should be optional unless your sdk ensures a value is always sent, even when user sets "null/undefined"

```suggestion
  if (options.nonce && !decoded.claims.nonce) {
```

File: src/jwt.ts [hunk 2]
```diff
@@ -53,37 +58,134 @@ export const decode = (token: string) => {
...
+    throw new Error(
+      'Nonce (nonce) claim must be a string present in the ID token'
+    );
+  }
+
   if (decoded.claims.nonce !== options.nonce) {
```
Comment (lbalmaceda) [inline src/jwt.ts]: this one should be wrapped within `if (options.nonce)` (see sample implementation for reference)

File: src/jwt.ts [hunk 3]
```diff
@@ -53,37 +58,134 @@ export const decode = (token: string) => {
...
+  }
+
+  if (options.max_age && !decoded.claims.auth_time) {
+    throw new Error(
+      'Authentication Time (auth_time) claim must be a number present in the ID token when Max Age (max_age) is specified'
+    );
+  }
+
+  /* istanbul ignore next */
+  if (!decoded.claims.exp) {
```
Comment (lbalmaceda) [inline src/jwt.ts]: check it's a number as well (applies to all date claims)

File: src/jwt.ts [hunk 4]
```diff
@@ -53,37 +58,134 @@ export const decode = (token: string) => {
...
+      'Issued At (iat) claim must be a number present in the ID token'
+    );
   }
 
   const now = new Date();
   const expDate = new Date(0);
   const iatDate = new Date(0);
   const nbfDate = new Date(0);
   const leeway = options.leeway || 60;
+  const authTimeDate = new Date(
```
Comment (lbalmaceda) [inline src/jwt.ts]: why is this one being instantiated like this and the rest are using the `setUTCSeconds`? Seems all should be using the latter

File: src/jwt.ts [hunk 5]
```diff
@@ -53,37 +58,134 @@ export const decode = (token: string) => {
...
+      `Issued At (iat) claim error in the ID token; current time (${now}) is before issued at time (${iatDate})`
+    );
   }
   if (typeof decoded.claims.nbf !== 'undefined' && now < nbfDate) {
-    throw new Error('token is not yet valid (invalid notBefore)');
+    throw new Error(
+      `Not Before time (nbf) claim in the ID token indicates that this token can't be used just yet. Currrent time (${now}) is before ${nbfDate}`
+    );
+  }
+  if (typeof decoded.claims.auth_time !== 'undefined' && now > authTimeDate) {
```
Comment (lbalmaceda) [inline src/jwt.ts]: Wouldn't it be more readable if you wrap this inside a presence-check if, like the one on line 140 ?

File: src/jwt.ts [hunk 6]
```diff
@@ -53,37 +58,134 @@ export const decode = (token: string) => {
 };
 
 export const verify = (options: JWTVerifyOptions) => {
+  if (!options.id_token) {
```
[the remark this answers is not in the document]
Comment (luisrudge) [reply src/jwt.ts]: this sdk always asks for id_token

File: src/jwt.ts (hunk 1 above)
Comment (luisrudge) [reply src/jwt.ts]: fixed

File: src/jwt.ts (hunk 5 above)
Comment (luisrudge) [reply src/jwt.ts]: but I only want to do the `now > authTimeDate` check if auth_time is present, so it's not the same check

File: src/jwt.ts (hunk 4 above)
Comment (luisrudge) [reply src/jwt.ts]: fixed

File: src/jwt.ts [hunk 7]
```diff
@@ -53,37 +60,136 @@ export const decode = (token: string) => {
...
+      'Expiration Time (exp) claim must be a number present in the ID token'
+    );
+  }
+  if (!isNumber(decoded.claims.iat)) {
+    throw new Error(
+      'Issued At (iat) claim must be a number present in the ID token'
+    );
   }
 
+  const leeway = options.leeway || 60;
```
Comment (esarafianou) [inline src/jwt.ts]: why do we set default leeway to 60 seconds? There is no mention about it in our docs.

Comment (gkwang) [reply src/jwt.ts]: We decided to set the default leeway to 60 across the board to keep the consistency.

Comment (esarafianou) [reply src/jwt.ts]: Makes sense. Let's include this piece of information here: https://auth0.github.io/auth0-spa-js/interfaces/auth0clientoptions.html#leeway

**Pull request merged**